closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -857,6 +863,63 @@ private void initConstraints() {
             }
         }
 
+        private void initProfileAttributes() {
+            if (debug != null) {
+                debug.println("\tLoading profile attributes.");
+            }
+
+            String pAttributes = parseProperty(
+                    securityProps.getProperty(profileID + ".attributes"));
+
+            if ((pAttributes == null) || pAttributes.isBlank()) {
+                if (debug != null) {
+                    debug.println("\tNo profile attributes.");
+                }
+                return;
+            }
+
+            try {
+                // Locate and parse providerattributes section.
+                pAttributes = pAttributes.replaceAll("\\s", "");
+                int startIndex = pAttributes.indexOf("<providerattributes>");
+                int endIndex = pAttributes.indexOf("</providerattributes>");
```

Comment:
I suggest that xml is not needed, nor is it very user-friendly in a properties file (i.e. `java.security`). The example in the description could be expressed more simply and clearly as:
```
RestrictedSecurity.NSS.140-2.attributes = \
    SunJGSS=sun.security.jgss.SunProvider, \
    SunSASL=com.sun.security.sasl.Provider
```
(possibly using a name other than `attributes`)

There are several references in comments to `providerattributes`; did you mean "provider attributes"?

Reply:
I considered this simply format(you suggested) before, but after discussed with the team, we think of a more general format. Right now, this XML tag “<providerattributes>” is used for the mapping between provider name and its class name. But, in the future, we can extend this "<restrictedsecurityattributes>" tag, to add other XML tags inside this “<restrictedsecurityattributes>” tag.

Reply:
Ok, but if you want to use xml, you should be using existing xml parsing classes.

Reply:
I take that back: java.base can't assume, for example, that the java.xml module is present and so can't use that code to parse xml.

Perhaps there's a better solution that doesn't require dealing with the "simple" name of a provider. Most of the uses of `isProviderAllowed(String)` could be switched to use `isProviderAllowed(Class)`. Could the remaining cases use the latter, or at least the (fully-qualfied) class name? If so, we don't need to discuss the format of the `attributes` property value.

Reply:
@keithc-ca 

If we only use provider class name to check allow or not-allow. That means, in java.security file, we have to use provider class name to configure the FIPS provider list. For example, when we want to add the SunSASL in the list, it will be like below:

`RestrictedSecurity.NSS.140-2.jce.provider.5=com.sun.security.sasl.Provider`

And then, in the `ProviderConfig` class, when loading this non-java.base provider, it will use `ServiceLoader` to load it, and due to “com.sun.security.sasl.Provider” is allowed (it can pass the `isProviderAllowed(Class)` check), so this provider class can be loaded. 

But after that, in `ProviderLoader.load()` method, it will use the provider name “SunSASL” to find the provider class object and return it [1]. But due to we did not use provider name in the java.security FIPS provider list, so the provider “SunSASL” can not be found and return (even this provider class is already loaded).

[1] - https://github.com/ibmruntimes/openj9-openjdk-jdk/blob/openj9/src/java.base/share/classes/sun/security/jca/ProviderConfig.java#L375

And also, in this `ProviderLoader` class, it do had another method called `legacyLoad()`. It is designed to load the provider by using the provider class name. But when I tried this method for "SunSASL", I get the below error:

> ProviderList: provider configuration: [com.sun.security.sasl.Provider]
> ProviderConfig: Loading legacy provider: com.sun.security.sasl.Provider
> ProviderConfig: Error loading legacy provider com.sun.security.sasl.Provider
> java.security.PrivilegedActionException: java.lang.IllegalAccessException: Class sun/security/jca/ProviderConfig$ProviderLoader$1(java.base) can not access class com/sun/security/sasl/Provider(java.security.sasl) because module java.security.sasl does not export package com/sun/security/sasl to module java.base

So, if the provider module did not export package to java.base module, then this `legacyLoad()` won't work for that provider class.

Reply:
@keithc-ca Any update or advice on this PR?

Reply:
Please update the description of this pull request: the discussion of unqualified names and XML are no longer relevant.

Reply:
Description updated.